tensorflow-ops/src/TensorFlow/Variable.hs

```diff
@@ -133,3 +140,53 @@ assignAdd = assignAdd' id
 assignAdd' :: (MonadBuild m, TensorType a)
     => OpParams -> Variable a -> Tensor v a -> m ControlNode
 assignAdd' params (Variable h _) v = CoreOps.assignAddVariableOp' params h v
+
+-- | Update '*var' according to the Adam algorithm.
+--
+-- lr_t <- learning_rate * sqrt(1 - beta2^t) / (1 - beta1^t)
+-- m_t <- beta1 * m_{t-1} + (1 - beta1) * g_t
+-- v_t <- beta2 * v_{t-1} + (1 - beta2) * g_t * g_t
+-- variable <- variable - lr_t * m_t / (sqrt(v_t) + epsilon)
+resourceApplyAdam :: forall v'4 v'5 v'6 v'7 v'8 v'9 v'10 t
```

Comment:
On second look, you can also probably drop the "forall" clauses from `resourceApplyAdam` and `resourceApplyAdam'` since these functions don't need ScopedTypeVariables.  (Up to you.